nemo_aligner/utils/distributed.py

```diff
@@ -304,6 +330,383 @@ def all_reduce_dict(dictionary, dtype=torch.float32, group=None, op=torch.distri
+            topk_logits, topk_token_ids = torch.topk(output_tensor, top_k)
+
+            return {
+                "topk_logits": topk_logits,
+                "topk_token_ids": topk_token_ids,
+                "log_sum_exp_logits": log_sum_exp_logits,
+            }
+
+        return output_tensor, fake_loss_func
+
+    seq_length = tokens.shape[1]
+    assert (
+        tokens.shape[0] % forward_micro_batch_size == 0
+    ), f"batch size {tokens.shape[0]} / forward_micro_batch_size {forward_micro_batch_size} is not divisible."
+    num_microbatches = int(tokens.shape[0] // forward_micro_batch_size)
+    data_iter = get_iterator_k_split(
+        {"tokens": tokens, "position_ids": position_ids, "attention_mask": attention_mask}, num_microbatches
+    )
+    fwd_bwd_function = get_forward_backward_func()
+    losses_reduced_per_micro_batch = fwd_bwd_function(
+        forward_step_func=fwd_output_and_loss_func,
+        data_iterator=data_iter,
+        model=model,
+        num_microbatches=num_microbatches,
+        forward_only=True,
+        seq_length=seq_length,
+        micro_batch_size=forward_micro_batch_size,
+        collect_non_loss_data=True,
+    )
+
+    if parallel_state.is_pipeline_last_stage():
+        topk_logits = torch.cat([item["topk_logits"] for item in losses_reduced_per_micro_batch], dim=0)
+        topk_token_ids = torch.cat([item["topk_token_ids"] for item in losses_reduced_per_micro_batch], dim=0)
+        log_sum_exp_logits = torch.cat([item["log_sum_exp_logits"] for item in losses_reduced_per_micro_batch], dim=0)
+    else:
+        topk_logits = None
+        topk_token_ids = None
+        log_sum_exp_logits = None
+
+    # broadcast it from last PP stage to everything else
+    if parallel_state.get_pipeline_model_parallel_world_size() > 1:
+        topk_logits = broadcast_tensor(
+            topk_logits,
+            parallel_state.get_pipeline_model_parallel_last_rank(),
+            parallel_state.get_pipeline_model_parallel_group(),
+            ndim=3,
+        )
+        topk_token_ids = broadcast_tensor(
+            topk_token_ids,
+            parallel_state.get_pipeline_model_parallel_last_rank(),
+            parallel_state.get_pipeline_model_parallel_group(),
+            ndim=3,
+        )
+        log_sum_exp_logits = broadcast_tensor(
+            log_sum_exp_logits,
+            parallel_state.get_pipeline_model_parallel_last_rank(),
+            parallel_state.get_pipeline_model_parallel_group(),
+            ndim=2,
+        )
+
+    return topk_logits, topk_token_ids, log_sum_exp_logits
+
+
+## how to handle conditionals?
+## separate implementations for forward and backward KL?
+class _TopKLogitsCrossEntropy(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        vocab_parallel_logits,
+        target_logits,
+        target_token_ids,
+        labels,
+        kd_loss_weight=1.0,
+        sft_loss_weight=0,
+        forward_kl=False,
```

Comment:
Can we pass in `kd_loss` as a string instead of `forward_kl` as a bool? It will ensure more flexibility in case if we need to add more loss functions.

Reply:
The reason I made it a boolean is because is because we need to save it for the backward pass, so I need to be able to convert it to a tensor. We could instead have an enum that represents the different loss variants. What do you think?

Reply:
Let's keep it this way then.